# Notebook 2: Summary Statistics and Boulder Weather Data 
***

In this notebook we'll: 
- Use Pandas to compute summary statistics on Boulder weather data 
- Figure out how summary statistics like mean and standard deviation change under transformations of the data


First, as always, we'll load Numpy and Pandas using their common aliases, np and pd. 

In [163]:
import numpy as np
import pandas as pd

The data we'll explore in this notebook concerns temperatures and other weather observations in Boulder County over the month of July 2017.  The data was obtained from the National Oceanic and Atmospheric Administration's [Climate.gov](https://www.climate.gov/) website.  You can find and download loads of climate-related data from NOAA [here](https://www.climate.gov/maps-data/datasets).   

The data is stored in a .csv file called clean_boulder_weather.csv.  If you've updated your local course repository then you can read the data into pandas from the given local path.  If you have not (e.g. if you're using Azure) then you can load the data into pandas by passing in the url to the raw data file on GitHub.  

In [164]:
# Path to the data
file_path = '/Users/Eli/Dropbox/FALL 2019 COURSES/CSCI 3022 Data Science/Data/clean_boulder_weather.csv'

# Load the data into a DataFrame 
df = pd.read_csv(file_path)

Take a look at the first 50 or so rows of the DataFrame using the head( ) method. 

In [165]:
df.head()

STATION                 NAME        DATE  PRCP  TMAX  TMIN
0  USW00094075  BOULDER 14 W, CO US  2017-07-01  0.00  68.0  31.0
1  USW00094075  BOULDER 14 W, CO US  2017-07-02  0.00  73.0  35.0
2  USW00094075  BOULDER 14 W, CO US  2017-07-03  0.00  68.0  46.0
3  USW00094075  BOULDER 14 W, CO US  2017-07-04  0.05  68.0  43.0
4  USW00094075  BOULDER 14 W, CO US  2017-07-05  0.01  73.0  40.0

From this you should see that each row in the DataFrame refers to a particular weather station / date combination.  The columns of the DataFrame are as follows: 

- **STATION**: The unique identification code for each weather station 
- **NAME**: The location / name of the weather station 
- **DATE**: The date of the observation 
- **PRCP**: The precipitation (in inches)
- **TMAX**: The daily maximum temperature (in Fahrenheit)
- **TMIN**: The daily minimum temperature (in Fahrenheit)

From the printed DataFrame above you can see that we actually have data from multiple weather stations.  To see how many, we can pass the **NAME** column (or the **STATION** column) into Python's set function. 

In [166]:
set(df["NAME"])

{'BOULDER 14 W, CO US',
 'BOULDER, CO US',
 'GROSS RESERVOIR, CO US',
 'NIWOT, CO US',
 'NORTHGLENN, CO US',
 'RALSTON RESERVOIR, CO US',
 'SUGARLOAF COLORADO, CO US'}

In [167]:
df["NAME"].unique()

array(['BOULDER 14 W, CO US', 'GROSS RESERVOIR, CO US',
       'SUGARLOAF COLORADO, CO US', 'NIWOT, CO US', 'BOULDER, CO US',
       'RALSTON RESERVOIR, CO US', 'NORTHGLENN, CO US'], dtype=object)

It looks like we have data from seven different weather stations.  For consistency, let's reduce the data to just the reports from the weather station in Niwot.  

### Exercise 1
***
Extract the rows of the DataFrame concerned with the Niwot weather station.  Store this data in a new DataFrame called dfNiwot. 

In [168]:
dfNiwot = df.loc[df["NAME"]=="NIWOT, CO US",:]  # TODO 
dfNiwot.head() 

STATION          NAME        DATE  PRCP  TMAX  TMIN
93  USS0005J42S  NIWOT, CO US  2017-07-01   0.0  69.0  32.0
94  USS0005J42S  NIWOT, CO US  2017-07-02   0.0  73.0  37.0
95  USS0005J42S  NIWOT, CO US  2017-07-03   0.0  68.0  47.0
96  USS0005J42S  NIWOT, CO US  2017-07-04   0.1  70.0  41.0
97  USS0005J42S  NIWOT, CO US  2017-07-05   0.0  74.0  40.0

### Exercise 2  
***

Pandas (and Numpy) have canned functions that compute each of the summary statistics discussed in lecture.  We'll use the .mean( ) function as an example.  All of these functions can be called either on a Pandas Series (i.e. a column of a DataFrame) or on an entire DataFrame at one time.  

For instance, the sample mean of the maximum daily temperature is given by 

In [169]:
dfNiwot["TMAX"].mean()

69.83870967741936

In [170]:
np.mean(dfNiwot["TMAX"])

69.83870967741936

Let's see what happens if we call .mean( ) on the entire DataFrame. 

In [171]:
dfNiwot.mean()

PRCP     0.061290
TMAX    69.838710
TMIN    43.548387
dtype: float64

In this case, Pandas returned a Series with the means of all of the **numerical** data in the DataFrame. 

The functions for the other summary statistics are as follows: 

\begin{array}{l|l}
\textrm{Function} & \textrm{Statistics} \\
\hline
\textrm{.var()} & \textrm{variance} \\
\textrm{.std()} & \textrm{standard deviation} \\
\textrm{.min()} & \textrm{minimum value} \\
\textrm{.max()} & \textrm{maximum value} \\
\textrm{.median()} & \textrm{value} \\
\textrm{.quantile(q)} & \textrm{quantile, where q is the desired percentage as a decimal} \\
\end{array}

Your job is to use these functions to compute the 5-number summary for the maximum daily temperature. 

var and std in PANDAS automatically does n-1 in denominator

but

var and std in NUMPY do not

In [172]:
dfNiwot["TMAX"].var()

31.606451612903232

In [173]:
np.var(dfNiwot["TMAX"], ddof=1)

31.606451612903232

In [174]:
FiveNS = "TMAX" #Since we are computing the 5 number summary for the max daily temp
minval = dfNiwot[FiveNS].min()
maxval = dfNiwot[FiveNS].max()
Q1 = dfNiwot[FiveNS].quantile(0.25)
Q2 = dfNiwot[FiveNS].quantile(0.5)
Q3 = dfNiwot[FiveNS].quantile(0.75)
print("5-Number Summary: {:.2f}    {:.2f}    {:.2f}    {:.2f}    {:.2f}".format(minval, Q1, Q2, Q3, maxval))
print(Q1)

5-Number Summary: 54.00    66.50    70.00    74.00    80.00
66.5


### Exercise 3 
***
It turns out that Pandas has a nice function called .describe( ) that will compute all of the standard summary statistics for you.  You can apply it either to a Pandas Series or to an entire DataFrame.  

Run the .describe( ) function on the **TMAX** column of your DataFrame, and check that the results agree with your computations from Exercise 2. 

In [175]:
dfNiwot["TMAX"].describe()

count    31.000000
mean     69.838710
std       5.621962
min      54.000000
25%      66.500000
50%      70.000000
75%      74.000000
max      80.000000
Name: TMAX, dtype: float64

### Exercise 4 
***
In this exercise we'll explore how the mean and the standard deviation change when we perform basic transformations on the data.  In particular, we're interested in what happens if we 

1. Add or subtract some value from every entry in the data set 
1. Multiply every entry in the data set by some value 

We know from above that the mean and standard deviation of the Niwot **TMAX** value are 69.83871 and 5.621962.  Experiment by adding and multiplying nice integer values with the **TMAX** column and then recomputing the statistics.  From your observations, can you guess how the mean and std dev change under these transformations? 

In [176]:
dfNiwotAdd = (dfNiwot[FiveNS].add(10)) # Std Dev stays the same when adding to all values

print(dfNiwotAdd.describe())

dfNiwotMult = (dfNiwot[FiveNS].multiply(10)) # Std Dev increases as datas becomes more spread when everything is multiplied

print(dfNiwotMult.describe())

count    31.000000
mean     79.838710
std       5.621962
min      64.000000
25%      76.500000
50%      80.000000
75%      84.000000
max      90.000000
Name: TMAX, dtype: float64
count     31.000000
mean     698.387097
std       56.219615
min      540.000000
25%      665.000000
50%      700.000000
75%      740.000000
max      800.000000
Name: TMAX, dtype: float64


See if you can prove that your guess works in general mathematically using the formulas for the two statistics: 

$$
\bar{x} = \frac{1}{n} \displaystyle\sum_{k=1}^n x_k \quad \quad \textrm{and} \quad \quad s = \sqrt{\frac{1}{n-1} \sum_{k=1}^n \left( x_k - \bar{x}\right)^2} 
$$

### Exercise 5 
***
OK, let's apply a common transformation to the **TMAX** and **TMIN** columns by converting the temperatures from Fahrenheit to Celsius.  Remember that the transformation is given by 

$$
\textrm{CELSIUS} = \frac{5}{9} (\textrm{FAHRENHEIT}-32) 
$$

First, use the Fahrenheit data in columns **TMAX** and **TMIN** to create Celsius columns in the Niwot DataFrame called **TMAX-C** and **TMIN-C**.

In [177]:
pd.options.mode.chained_assignment = None # Essentially ignores the settingwithcopy warning
# Remember iloc is for unknown indexes in a row or column
dfNiwot.loc[:,"TMAX-C"] = (5/9)*(dfNiwot.loc[:, FiveNS] - 32)

dfNiwot.loc[:,"TMIN-C"] = (5/9)*(dfNiwot.loc[:, "TMIN"] - 32)

In [178]:
dfNiwot.head()

STATION          NAME        DATE  PRCP  TMAX  TMIN     TMAX-C  \
93  USS0005J42S  NIWOT, CO US  2017-07-01   0.0  69.0  32.0  20.555556   
94  USS0005J42S  NIWOT, CO US  2017-07-02   0.0  73.0  37.0  22.777778   
95  USS0005J42S  NIWOT, CO US  2017-07-03   0.0  68.0  47.0  20.000000   
96  USS0005J42S  NIWOT, CO US  2017-07-04   0.1  70.0  41.0  21.111111   
97  USS0005J42S  NIWOT, CO US  2017-07-05   0.0  74.0  40.0  23.333333   

      TMIN-C  
93  0.000000  
94  2.777778  
95  8.333333  
96  5.000000  
97  4.444444

**Note**: You may get a SetWithCopyWarning even if you're correctly using loc.  If you click the link to the documentation that comes with the warning you'll find a nice statement that says that SetWithCopyWarning can give false positives sometimes. 

Based on the stuff we proved in **Exercise 4**, what do you expect the mean and the standard deviation of the daily maximum temperature to be in Celsius? 

Once you've made your guess, see if you're right by applying the .describe( ) method to **TMAX-C** and **TMIN-C**. 

In [179]:
print(dfNiwot["TMAX-C"].describe())

print(dfNiwot["TMIN-C"].describe())

count    31.000000
mean     21.021505
std       3.123312
min      12.222222
25%      19.166667
50%      21.111111
75%      23.333333
max      26.666667
Name: TMAX-C, dtype: float64
count    31.000000
mean      6.415771
std       2.235326
min       0.000000
25%       5.000000
50%       6.666667
75%       8.333333
max      11.111111
Name: TMIN-C, dtype: float64


### Exercise 6 
***

Compute the daily temperature range (max minus min) for each row in the Niwot DataFrame and store it in a column called **TDIFF**.  Then answer these questions.  

- What is the mean temperature difference over the month of July? 
- What is the difference between the means of the max and min daily temperatures? 
- Do you see a relationship between these two quantities?  If so, can you prove that it's always the case for mean difference and difference of means? 

In [182]:
dfNiwot.loc[:,"TDIFF"] = dfNiwot["TMAX"] - dfNiwot["TMIN"] # Temp Range

dfNiwot.head()

STATION          NAME        DATE  PRCP  TMAX  TMIN     TMAX-C  \
93  USS0005J42S  NIWOT, CO US  2017-07-01   0.0  69.0  32.0  20.555556   
94  USS0005J42S  NIWOT, CO US  2017-07-02   0.0  73.0  37.0  22.777778   
95  USS0005J42S  NIWOT, CO US  2017-07-03   0.0  68.0  47.0  20.000000   
96  USS0005J42S  NIWOT, CO US  2017-07-04   0.1  70.0  41.0  21.111111   
97  USS0005J42S  NIWOT, CO US  2017-07-05   0.0  74.0  40.0  23.333333   

      TMIN-C  TDIFF Month  
93  0.000000   37.0    07  
94  2.777778   36.0    07  
95  8.333333   21.0    07  
96  5.000000   29.0    07  
97  4.444444   34.0    07

1. The mean temperature difference over the month of July is 26.29032258064516

2. The difference of the means between the max and min daily temperatures is 

3. The relationship between these two quantities is that they are the same since there are only July dates in dfNiwot.

In [186]:
dfDate = dfNiwot["DATE"].str.split("-", n = 2 , expand = True)
dfNiwot["Month"] = dfDate[1]
print("The mean temp difference over the month of July is %5.2f" % dfNiwot.loc[:,"TDIFF"].mean())# Mean temp difference for the month of July

maxMinDiff = dfNiwot.loc[:,"TMAX"].mean() - dfNiwot.loc[:,"TMIN"].mean() # Mean temp difference of daily temperatures
print("The mean temp difference for max and min daily temps is %5.2f" % maxMinDiff)

The mean temp difference over the month of July is 26.29
The mean temp difference for max and min daily temps is 26.29
